packages/react-router/modules/matchPath.js

```diff
@@ -44,15 +50,48 @@ const matchPath = (pathname, path, options = {}) => {
   if (exact && !isExact)
     return null
 
+  const matchParams = keys.reduce((memo, key, index) => {
+    memo[key.name] = values[index]
+    return memo
+  }, {})
+
+  // merge parent match's params for relative paths
+  // this allows resolving using parent match's url instead of path
+  const params = absolute
+    ? matchParams
+    : Object.assign({}, parentMatch && parentMatch.params, matchParams)
+
   return {
     path, // the path pattern used to match
     url: path === '/' && url === '' ? '/' : url, // the matched portion of the URL
     isExact, // whether or not we matched exactly
-    params: keys.reduce((memo, key, index) => {
-      memo[key.name] = values[index]
-      return memo
-    }, {})
+    params
+  }
+}
+
+const resolvePath = (pathname, base) => {
+  if (pathname === undefined || isAbsolute(pathname)) {
+    return pathname
+  }
+
+  if (!base) {
+    base = '/'
+  }
+
+  if (pathname === '') {
+    return base
+  } else {
+    return `${withTrailingSlash(base)}${pathname}`
   }
 }
 
+const isAbsolute = pathname => !!(pathname && pathname.charAt(0) === '/')
+
+const withTrailingSlash = pathname =>
```

Comment:
Let's call this `addTrailingSlash`. At some point, it may make sense for history and router to use the same bag of functions for this stuff.